toolchain/docs/idioms.md

```diff
@@ -197,39 +208,31 @@ The presence of specific fields in a struct with a specified type is detected
 using the following idiom:
 
 ```cpp
-template <typename T, typename = FieldType T::*>
-constexpr bool HasField = false;
-template <typename T>
-constexpr bool HasField<T, decltype(&T::field)> = true;
-```
-
-This is intended to check the same property as the following concept, which we
-can't use because we currently need to compile in C++17 mode:
-
-```cpp
+// HasField<T> is true if T has a `U field` field of type FieldType.
 template <typename T> concept HasField = requires (T x) {
-  { x.field } -> std::same_as<FieldType>;
+  { &x.field } -> std::same_as<FieldType T::*>;
 };
 ```
 
+See `HasKindMemberAsField` in
+[`toolchain/sem_ir/typed_insts.h`](/toolchain/sem_ir/typed_insts.h) for an
+example.
+
 To detect a field with a specific name with a type derived from a specified base
 type, use this idiom:
 
 ```cpp
-// HasField<T> is true if T has a `U field` field,
-// where `U` extends `BaseClass`.
-template <typename T, bool Enabled = true>
-inline constexpr bool HasField = false;
-template <typename T>
-inline constexpr bool HasField<
-    T, bool(std::is_base_of_v<BaseClass, decltype(T::field)>)> = true;
-```
+namespace Internal {
 
-The equivalent concept is:
+template <typename T, typename From> concept FieldDerivedFrom =
+    std::derived_from<std::remove_cvref_t<T>, From>;
 
-```cpp
+}
+
+// HasField<T> is true if T has a `U field` field, where `U` extends
+// `BaseClass`.
 template <typename T> concept HasField = requires (T x) {
-  { x.field } -> std::derived_from<BaseClass>;
+  { x.field } -> Internal::FieldDerivedFrom<BaseClass>;
 };
 ```
```

Comment:
If we're not using this idiom, maybe we should remove it from the documentation for now.

Reply:
Yeah, especially since it needs a helper. SGTM